ob1k-core/src/main/java/com/outbrain/ob1k/server/builder/ServiceRegisterBuilder.java

```diff
@@ -0,0 +1,57 @@
+package com.outbrain.ob1k.server.builder;
+
+import com.outbrain.ob1k.Service;
+import com.outbrain.ob1k.common.filters.ServiceFilter;
+import com.outbrain.ob1k.server.builder.ServiceBindBuilder.ServiceBindBuilderSection;
+
+/**
+ * portion of the ServerBuilder methods used for service registration.
+ *
+ * ports, timeouts, listeners etc.
+ *
+ *
+ * @param <B> ability to extend this builder
+ */
+public class ServiceRegisterBuilder<B extends ServiceRegisterBuilder<B>> {
+
+  /**
+   * This non generic interface is used to bypass a Java 8 lambda compiler issue
+   * where the compiler fails to infer the lambda argument if those are generic type in a method
+   * for a genericized class.
+   */
+  public interface ServiceRegisterBuilderSection extends BuilderSection<ServiceRegisterBuilder> {}
+
+  private static final NoOpBindSection NO_OP = new NoOpBindSection();
+
+  private final ServerBuilderState state;
+  private final ServiceBindBuilder bindBuilder;
+
+  public ServiceRegisterBuilder(final ServerBuilderState state) {
+    this.state = state;
+    this.bindBuilder = new ServiceBindBuilder(state);
+
+  }
+
+  public B register(final Service service, final String path, final ServiceFilter... filters) {
+    return register(service, path, NO_OP, filters);
+  }
+
+  public B register(final Service service, final String path, final ServiceBindBuilderSection bindSection, final ServiceFilter... filters) {
+    state.addServiceDescriptor(service, path, filters);
+    bindSection.apply(bindBuilder);
+    return self();
+  }
+
+  @SuppressWarnings("unchecked")
+  private B self() {
+    return (B) this;
+  }
+
+  private static class NoOpBindSection implements ServiceBindBuilderSection {
```

Comment:
I prefer this approach.
It creates code which less complex. It has less possible flows, normal flows don't need to check for null and I don't have a magic value (null) infused with a meaning (i.e do nothing = no op), I actually have the meaning very explicitly in the inner class name.
I only create one single instance of this class so no memory penalty either.

The most important thing: If a user accidently inserts null as an argument instead of a 'bindSection' it is not going to just work! It actually is going to blow up (which is what you want).